lib/player/db/postgres.go

```diff
@@ -0,0 +1,138 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package db
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gravitational/teleport/api/types/events"
+)
+
+// PostgresTranslator is responsible for converting PostgreSQL session recording
+// events into their text representation.
+type PostgresTranslator struct {
+	sessionStartAt                 time.Time
+	preparedStatements             map[string]string
+	lastPreparedStatement          string
+	lastPreparedStatementArguments []string
+	expectingResult                bool
+}
+
+// NewPostgresTranslator constructs a new instance of PostgreSQL translator.
+func NewPostgresTranslator() *PostgresTranslator {
+	return &PostgresTranslator{
+		preparedStatements: make(map[string]string),
+		expectingResult:    false,
+	}
+}
+
+// TranslateEvent converts PostgreSQL audit events into session print events.
+func (p *PostgresTranslator) TranslateEvent(evt events.AuditEvent) *events.SessionPrint {
+	switch e := evt.(type) {
+	case *events.DatabaseSessionQuery:
+		p.expectingResult = true
+		return p.generatePrintEvent(e.Metadata, p.generateCommandPrint(e.DatabaseMetadata, e.DatabaseQuery))
+	case *events.DatabaseSessionCommandResult:
+		if p.expectingResult {
+			p.expectingResult = false
+			return p.generatePrintEvent(e.Metadata, p.generateResultPrint(e))
+		}
+	case *events.PostgresParse:
+		p.expectingResult = false
+		p.preparedStatements[e.StatementName] = e.Query
+	case *events.PostgresBind:
+		p.expectingResult = false
+		p.lastPreparedStatement = e.StatementName
+		p.lastPreparedStatementArguments = e.Parameters
+	case *events.PostgresExecute:
+		printEvent := p.generatePreparedStatementPrint(e.Metadata, e.DatabaseMetadata)
+		p.expectingResult = printEvent != nil
+		p.lastPreparedStatement = ""
+		p.lastPreparedStatementArguments = nil
+		return printEvent
+	case *events.PostgresFunctionCall:
+		p.expectingResult = false
+	case *events.DatabaseSessionStart:
+		p.sessionStartAt = e.Time
+	case *events.DatabaseSessionEnd:
+	}
+
+	return nil
+}
+
+const lineBreak = "\r\n"
+
+func (p *PostgresTranslator) generateCommandPrint(metadata events.DatabaseMetadata, command string) string {
+	return lineBreak + fmt.Sprintf("%s=> %s", metadata.DatabaseName, command) + lineBreak
+}
+
+func (p *PostgresTranslator) generatePreparedStatementPrint(metadata events.Metadata, databaseMetadata events.DatabaseMetadata) *events.SessionPrint {
+	// If there wasn't a bind before the execute, we don't know which prepared
+	// statement or arguments were executed. In this case, we skip the prepared
+	// statement, ignore any possible result too.
+	if p.lastPreparedStatement == "" {
```

Comment:
Assuming this shouldn't happen right? maybe log a warning.

Reply:
It can be a mistake or a malformed request (I have yet to see a case where this happens to return success). Initially, I returned the information as an error (usually, the execution comes with a result event). Something like:

```
postgres=> Execute unknown prepared statement;
ERROR: prepared statement \"test\" does not exist (SQLSTATE 26000)
```

But I removed it since I couldn't find a way to present it to users. If you have any suggestions, let me know.